src/tests/fuzzing/fuzz_oid.c

```diff
@@ -0,0 +1,75 @@
+/* -*- mode: c; c-basic-offset: 4; indent-tabs-mode: nil -*- */
+/* tests/fuzzing/fuzz_oid.c */
+/*
+ * Copyright (C) 2024 by Arjun. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ *
+ * * Redistributions of source code must retain the above copyright
+ *   notice, this list of conditions and the following disclaimer.
+ *
+ * * Redistributions in binary form must reproduce the above copyright
+ *   notice, this list of conditions and the following disclaimer in
+ *   the documentation and/or other materials provided with the
+ *   distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ * FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ * COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT,
+ * INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+ * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
+ * STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED
+ * OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/*
+ * Fuzzing harness implementation for gss_oid_to_str, gss_str_to_oid.
+ */
+
+#include "autoconf.h"
+#include <k5-int.h>
+#include <gssapi/gssapi_krb5.h>
+
+#define kMinInputLength 2
+#define kMaxInputLength 1024
+
+extern int LLVMFuzzerTestOneInput(const uint8_t *data, size_t size);
+
+int
+LLVMFuzzerTestOneInput(const uint8_t *data, size_t size)
+{
+    OM_uint32 minor;
+    gss_buffer_desc buf;
+    gss_OID oid;
+    gss_OID_desc oid_desc;
+
+    if (size < kMinInputLength || size > kMaxInputLength)
+        return 0;
+
+    oid_desc.elements = (void *)data;
+    oid_desc.length = size;
+
+    gss_oid_to_str(&minor, &oid_desc, &buf);
+    gss_release_buffer(&minor, &buf);
+
+    buf.value = (void *)data;
+    buf.length = size;
+
+    gss_str_to_oid(&minor, &buf, &oid);
+
+    // gss_release_oid() is buggy.
```

Comment:
`gss_release_oid()` does call the `gssint_mechglue_initialize_library`->`gssint_mechglue_init`.
but it never calls the `gssint_mechglue_fini`, so there is a global variable memory leak.

### Foot note

I don't know how to fix this. I tried, though.

Reply:
What mechanism is detecting the memory leak?  If at process exit there is allocated memory that is reachable from global variables, that shouldn't be considered an error (for instance, it happens in libc if you ever call dlopen() with RTLD_NODELETE).

Reply:
### compile and run
```bash
export CC=clang
export CXX=clang++
export CFLAGS="-g -DFUZZING_BUILD_MODE_UNSAFE_FOR_PRODUCTION -fsanitize=address -fsanitize=fuzzer-no-link"
export CXXFLAGS="-g -DFUZZING_BUILD_MODE_UNSAFE_FOR_PRODUCTION -fsanitize=address -fsanitize=fuzzer-no-link"
export LIB_FUZZING_ENGINE="-fsanitize=fuzzer"

mkdir -p build
cd build
./../src/configure CFLAGS="-fcommon $CFLAGS" CXXFLAGS="-fcommon $CXXFLAGS" --enable-static --disable-shared --enable-ossfuzz

# fuzz_attrset will fail here, but that's okay
make -j$(nproc) --ignore-errors
echo "Brrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrr" > save.raw
./tests/fuzzing/fuzz_oid save.raw
```

### asan-logs
```c
INFO: Running with entropic power schedule (0xFF, 100).
INFO: Seed: 2116037642
INFO: Loaded 1 modules   (24200 inline 8-bit counters): 24200 [0xa8ec10, 0xa94a98), 
INFO: Loaded 1 PC tables (24200 PCs): 24200 [0x9d5288,0xa33b08), 
./tests/fuzzing/fuzz_oid: Running 1 inputs 1 time(s) each.
Running: save.raw
Executed save.raw in 0 ms
***
*** NOTE: fuzzing was not performed, you have only
***       executed the target code on a fixed set of inputs.
***

=================================================================
==310003==ERROR: LeakSanitizer: detected memory leaks

Direct leak of 32 byte(s) in 1 object(s) allocated from:
    #0 0x500ee0 in realloc (/home/brrr/krb5/build/tests/fuzzing/fuzz_oid+0x500ee0) (BuildId: 1dac8079cd2e6b709ecef5bef948c6033b347e46)
    #1 0x611084 in mecherrmap__pairarray_grow /home/brrr/krb5/build/lib/gssapi/generic/./errmap.h:106:11
    #2 0x60ff2e in mecherrmap_add /home/brrr/krb5/build/lib/gssapi/generic/./errmap.h:205:11
    #3 0x60e8f4 in gssint_mecherrmap_map /home/brrr/krb5/build/lib/gssapi/generic/../../.././../src/lib/gssapi/generic/util_errmap.c:223:11
    #4 0x6100fa in gssint_mecherrmap_map_errcode /home/brrr/krb5/build/lib/gssapi/generic/../../.././../src/lib/gssapi/generic/util_errmap.c:244:12
    #5 0x56365d in gss_str_to_oid /home/brrr/krb5/build/lib/gssapi/mechglue/../../.././../src/lib/gssapi/mechglue/g_oid_ops.c:77:2
    #6 0x541883 in LLVMFuzzerTestOneInput /home/brrr/krb5/build/tests/fuzzing/../.././../src/tests/fuzzing/fuzz_oid.c:65:5
    #7 0x44b994 in fuzzer::Fuzzer::ExecuteCallback(unsigned char const*, unsigned long) (/home/brrr/krb5/build/tests/fuzzing/fuzz_oid+0x44b994) (BuildId: 1dac8079cd2e6b709ecef5bef948c6033b347e46)
    #8 0x4347ea in fuzzer::RunOneTest(fuzzer::Fuzzer*, char const*, unsigned long) (/home/brrr/krb5/build/tests/fuzzing/fuzz_oid+0x4347ea) (BuildId: 1dac8079cd2e6b709ecef5bef948c6033b347e46)
    #9 0x43a130 in fuzzer::FuzzerDriver(int*, char***, int (*)(unsigned char const*, unsigned long)) (/home/brrr/krb5/build/tests/fuzzing/fuzz_oid+0x43a130) (BuildId: 1dac8079cd2e6b709ecef5bef948c6033b347e46)
    #10 0x465ea6 in main (/home/brrr/krb5/build/tests/fuzzing/fuzz_oid+0x465ea6) (BuildId: 1dac8079cd2e6b709ecef5bef948c6033b347e46)
    #11 0x7f938100b087 in __libc_start_call_main (/lib64/libc.so.6+0x2a087) (BuildId: 77c77fee058b19c6f001cf2cb0371ce3b8341211)
    #12 0x7f938100b14a in __libc_start_main@GLIBC_2.2.5 (/lib64/libc.so.6+0x2a14a) (BuildId: 77c77fee058b19c6f001cf2cb0371ce3b8341211)
    #13 0x42ee04 in _start (/home/brrr/krb5/build/tests/fuzzing/fuzz_oid+0x42ee04) (BuildId: 1dac8079cd2e6b709ecef5bef948c6033b347e46)

SUMMARY: AddressSanitizer: 32 byte(s) leaked in 1 allocation(s).
```

Reply:
>  If at process exit there is allocated memory that is reachable from global variables, that shouldn't be considered an error (for instance, it happens in libc if you ever call dlopen() with RTLD_NODELETE).

https://github.com/krb5/krb5/blob/038793c3083f44c4fb62626c12f80c80147029cf/src/lib/gssapi/generic/util_errmap.c#L133-L135

https://github.com/krb5/krb5/blob/038793c3083f44c4fb62626c12f80c80147029cf/src/lib/win_glue.c#L364-L366


`static mecherrmap m;` is the variable where memory is allocated.
`gssint_mechglue_fini` is never called except for `win` `DLL_SHUTDOWN`.

Reply:
I see.  The problem is not that the finalizer isn't running (that's deliberate, see 1bfcf572241a4ec0e44e609e5c6b7c0b11b08eea) or that asan is reporting still-reachable allocations as a leak.  The problem is that in this program, gss_str_to_oid() or gss_oid_to_str() can add an entry the error map _before_ the initializer runs.  When gss_release_oid() triggers the initializer, it re-initializes the error map, leaking the entry contained within.  So even if the finalizer was run there would be a (properly detected) leak.

I think the fix is to make gssint_mecherrmap_map() ensure that the library is initialized.  I tested a proof of concept of that and it squashes the error as expected, but I need to figure out whether there's a more elegant way of doing it.

Reply:
PR #1383 should address this problem.

Reply:
> The problem is that in this program, gss_str_to_oid() or gss_oid_to_str() can add an entry the error map before the initializer runs. When gss_release_oid() triggers the initializer, it re-initializes the error map, leaking the entry contained within.

> PR https://github.com/krb5/krb5/pull/1383 should address this problem.

This work,  thanks.

Reply:
I had a little concern over this call `gss_release_oid()->gssint_mechglue_initialize_library()`.
But `CALL_INIT_FUNCTION` is handling it great.

https://github.com/krb5/krb5/blob/038793c3083f44c4fb62626c12f80c80147029cf/src/lib/gssapi/mechglue/g_initialize.c#L182-L185